cranelift/codegen/src/remove_constant_phis.rs

```diff
@@ -0,0 +1,401 @@
+//
+// Work proceeds in three phases.
+//
+// * Phase 1: examine all instructions.  For each block, make up a useful
+//   grab-bag of information, `BlockSummary`, that summarises the block's
+//   formals and jump/branch instruction.  This is used by Phases 2 and 3.
+//
+// * Phase 2: for each value in Group A, try to find a single Group B value
+//   that flows to it.  This is done using a classical iterative forward
+//   dataflow analysis over a simple constant-propagation style lattice.  It
+//   converges quickly in practice -- I have seen at most 4 iterations.  This
+//   is relatively cheap because the iteration is done over the
+//   `BlockSummary`s, and does not visit each instruction.  The resulting
+//   fixed point is stored in a `SolverState`.
+//
+// * Phase 3: using the `SolverState` and `BlockSummary`, edit the function to
+//   remove redundant formals and actuals, and to insert suitable renames.
+//
+// Note that the effectiveness of the analysis depends on on the fact that
+// there are no copy instructions in Cranelift's IR.  If there were, the
+// computation of `actual_absval` in Phase 2 would have to be extended to
+// chase through such copies.
+//
+// For large functions, the analysis cost using the new AArch64 backend is about
+// 0.6% of the non-optimising compile time, as measured by instruction counts.
+// This transformation usually pays for itself several times over, though, by
+// reducing the isel/regalloc cost downstream.  Gains of up to 7% have been
+// seen for large functions.
+
+// The `Value`s (Group B) that can flow to a formal parameter (Group A).
+#[derive(Clone, Copy, Debug)]
+enum AbstractValue {
+    // Two or more values flow to this formal.
+    Many,
+    // Exactly one value, as stated, flows to this formal.  The `Value`s that
+    // can appear here are exactly: `Value`s defined by `Inst`s, plus the
+    // `Value`s defined by the formals of the entry block.  Note that this is
+    // exactly the set of `Value`s that are *not* tracked in the solver below
+    // (see `SolverState`).
+    One(Value /*Group B*/),
+    // No value flows to this formal.
+    None,
+}
+
+impl AbstractValue {
+    fn join(self, other: AbstractValue) -> AbstractValue {
+        match (self, other) {
+            // Joining with `None` has no effect
+            (AbstractValue::None, p2) => p2,
+            (p1, AbstractValue::None) => p1,
+            // Joining with `Many` produces `Many`
+            (AbstractValue::Many, _p2) => AbstractValue::Many,
+            (_p1, AbstractValue::Many) => AbstractValue::Many,
+            // The only interesting case
+            (AbstractValue::One(v1), AbstractValue::One(v2)) => {
+                if v1 == v2 {
+                    AbstractValue::One(v1)
+                } else {
+                    AbstractValue::Many
+                }
+            }
+        }
+    }
+    fn equals(self, other: AbstractValue) -> bool {
```

Comment:
(detached from other thread because this line changed -- related to `Eq` discussion)

Yup, I'd prefer to stay idiomatic here and remove `fn equals`, using the derived `Eq` instead -- if the derived code isn't doing the right thing then we have bigger problems (and others who maintain the code in the future might expect to use `==` anyway).